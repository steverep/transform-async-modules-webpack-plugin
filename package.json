{
  "name": "transform-async-modules-webpack-plugin",
  "description": "A Webpack plugin to transpile async module output using Babel. Allows transpiling top level await to ES5.",
  "version": "1.1.1",
  "license": "MIT",
  "homepage": "https://github.com/steverep/transform-async-modules-webpack-plugin",
  "keywords": [
    "webpack plugin",
    "async modules",
    "top level await",
    "ES5",
    "babel"
  ],
  "author": {
    "name": "Steve Repsher"
  },
  "repository": {
    "url": "https://github.com/steverep/transform-async-modules-webpack-plugin.git",
    "type": "git"
  },
  "bugs": {
    "url": "https://github.com/steverep/transform-async-modules-webpack-plugin/issues"
  },
  "funding": {
    "type": "github",
    "url": "https://github.com/sponsors/steverep"
  },
  "packageManager": "yarn@4.9.1",
  "engines": {
    "node": ">=16"
  },
  "type": "module",
  "files": [
    "dist/",
    "!dist/*.tsbuildinfo"
  ],
  "exports": "./dist/index.cjs",
  "types": "./dist/index.d.cts",
  "scripts": {
    "lint:js": "eslint --cache-location=node_modules/.cache/eslint/.eslintcache .",
    "lint:format": "prettier --cache --check .",
    "lint": "yarn dedupe --check && yarn run lint:js && yarn run lint:format",
    "fix": "yarn run lint:js --fix",
    "format": "prettier --cache --write .",
    "build": "tsc --build --verbose",
    "test": "mocha --config mocha.config.cjs",
    "snapshot": "CHAI_JEST_SNAPSHOT_UPDATE_ALL=trueÂ yarn run test",
    "clean": "tsc --build --clean && rm -rf dist test/dist",
    "prepublish": "yarn run lint",
    "prepack": "yarn run build",
    "postpack": "yarn run clean"
  },
  "dependencies": {
    "@babel/core": "^7.13.0",
    "@babel/plugin-transform-runtime": "^7.13.0",
    "@babel/preset-env": "^7.13.0"
  },
  "peerDependencies": {
    "@babel/core": "^7.13.0",
    "@babel/plugin-transform-runtime": "^7.13.0",
    "@babel/preset-env": "^7.13.0",
    "@babel/runtime": "^7.26.10",
    "webpack": "^5.94.0"
  },
  "devDependencies": {
    "@babel/runtime": "7.27.1",
    "@eslint/js": "9.27.0",
    "@types/babel__core": "7.20.5",
    "@types/babel__plugin-transform-runtime": "7.9.5",
    "@types/babel__preset-env": "7.10.0",
    "@types/chai": "5.2.2",
    "@types/chai-jest-snapshot": "1.3.8",
    "@types/eslint-config-prettier": "6.11.3",
    "@types/eslint__js": "8.42.3",
    "@types/mocha": "10.0.10",
    "@types/node": "22.15.30",
    "chai": "5.2.0",
    "chai-jest-snapshot": "2.0.0",
    "eslint": "9.27.0",
    "eslint-config-prettier": "10.1.5",
    "globals": "15.15.0",
    "mocha": "11.5.0",
    "prettier": "3.5.3",
    "ts-node": "10.9.2",
    "typescript": "5.8.3",
    "typescript-eslint": "8.32.1",
    "webpack": "5.99.8"
  }
}
