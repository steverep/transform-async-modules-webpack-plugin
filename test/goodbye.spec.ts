import { expect } from "chai";
import { readFile } from "node:fs/promises";
import { createRequire } from "node:module";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { TransformAsyncModulesPlugin } from "transform-async-modules-webpack-plugin";
import webpack from "webpack";

type GoodbyeEntry = typeof import("./src/simple.js");

const require = createRequire(import.meta.url);
const { devDependencies } =
  require("../package.json") as typeof import("../package.json");
const context = fileURLToPath(new URL(".", import.meta.url));
const ENTRY_NAMES = ["simple", "chained"] as const;
const CHUNK_CHECKS = { goodbye: 1, parent: 2 } as const;

const numMatchesTest = (file: string, pattern: RegExp, nMatches: number) =>
  async function () {
    const code = await readFile(file, "utf-8");
    expect(code.match(new RegExp(pattern, "g"))?.length).to.equal(nMatches);
  };

for (const runtime of [false, { version: devDependencies["@babel/runtime"] }]) {
  const dashRuntime = runtime ? "-runtime" : "";
  for (const nodeVersion of [8, 6, 0.12]) {
    describe(`Goodbye delay with node ${nodeVersion} target${runtime ? " using runtime" : ""}`, function () {
      const outDir = join(context, "dist", `v${nodeVersion}${dashRuntime}`);
      const configName = `goodbye-v${nodeVersion}${dashRuntime}`;
      let log: string | undefined;

      before("Run Webpack", function (done) {
        this.timeout("5s");
        webpack(
          {
            name: configName,
            context,
            entry: Object.fromEntries(
              ENTRY_NAMES.map((name) => [name, `./src/${name}.js`]),
            ),
            output: {
              clean: true,
              path: outDir,
              filename: "[name].cjs",
              library: { type: "commonjs-static" },
            },
            target: `async-node${nodeVersion}`,
            plugins: [
              new TransformAsyncModulesPlugin({
                browserslistConfigFile: false,
                targets: `node ${nodeVersion}`,
                runtime,
              }),
            ],
            devtool: "source-map",
            mode: "development",
            node: false,
            // infrastructureLogging: { debug: /TransformAsyncModulesPlugin/ },
          },
          (err, stats) => {
            log = stats?.toString({
              preset: "errors-warnings",
              loggingDebug: [/TransformAsyncModulesPlugin/],
              errorDetails: true,
            });
            done(err);
          },
        );
      });

      it("Compiles without errors or warnings", async function () {
        expect(log).to.match(
          new RegExp(`${configName} compiled successfully$`),
          log,
        );
      });

      for (const [name, nMatches] of Object.entries(CHUNK_CHECKS)) {
        const outChunk = join(outDir, `src_${name}_js.cjs`);
        if (nodeVersion < 7.6) {
          it(
            `Transpiles the async function to a generator in the ${name} chunk`,
            numMatchesTest(
              outChunk,
              runtime
                ? /var _asyncToGenerator = __webpack_require__/
                : /function _asyncToGenerator\(fn\) {/,
              nMatches,
            ),
          );
        } else {
          it(
            `Contains async arrow function in the ${name} chunk`,
            numMatchesTest(
              outChunk,
              / async \(__webpack_handle_async_dependencies__, __webpack_async_result__\) => {/,
              nMatches,
            ),
          );
        }
        if (nodeVersion < 4) {
          it(
            `Uses the regenerator runtime in the ${name} chunk`,
            numMatchesTest(
              outChunk,
              runtime
                ? /var _regeneratorRuntime = __webpack_require__/
                : / function _regeneratorRuntime\(\) {/,
              nMatches,
            ),
          );
        }
      }

      for (const name of ENTRY_NAMES) {
        const outEntry = join(outDir, `${name}.cjs`);
        it(`Awaits with correct phrase and delay for ${name} entry`, async function () {
          const { goodbye } = (await import(outEntry)) as GoodbyeEntry;
          const { PHRASE, DELAY, awaitedPhrase, awaitedDelay } = await goodbye;
          expect(awaitedPhrase).to.be.a("string").which.equals(PHRASE);
          expect(awaitedDelay)
            .to.be.a("number")
            .which.is.at.least(0.99 * DELAY);
        });
      }
    });
  }
}
