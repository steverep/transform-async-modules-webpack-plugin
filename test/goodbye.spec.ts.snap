// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Goodbye delay with node 0.12 target Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v0.12 compiled successfully"
`;

exports[`Goodbye delay with node 0.12 target using runtime Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Adding dependencies for top level await parsed in ./src/goodbye.js
    Adding dependencies for top level await parsed in ./src/another.js
    Adding dependencies for 2 more modules:
    ./src/parent.js
    ./src/with2.js
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v0.12-runtime compiled successfully"
`;

exports[`Goodbye delay with node 6 target Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v6 compiled successfully"
`;

exports[`Goodbye delay with node 6 target using runtime Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Adding dependencies for top level await parsed in ./src/goodbye.js
    Adding dependencies for top level await parsed in ./src/another.js
    Adding dependencies for 2 more modules:
    ./src/parent.js
    ./src/with2.js
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v6-runtime compiled successfully"
`;

exports[`Goodbye delay with node 8 target Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v8 compiled successfully"
`;

exports[`Goodbye delay with node 8 target using runtime Compiles with correct debug log 1`] = `
"DEBUG LOG from TransformAsyncModulesPlugin
    Transforming module ./src/goodbye.js for chunk src_goodbye_js
    Transforming module ./src/goodbye.js for chunk src_parent_js
    Transforming module ./src/parent.js for chunk src_parent_js
    Transforming module ./src/another.js for chunk src_with2_js
    Transforming module ./src/goodbye.js for chunk src_with2_js
    Transforming module ./src/with2.js for chunk src_with2_js

goodbye-v8-runtime compiled successfully"
`;
